[Arm64] Enable SIMD

sdmaclea opened the pull request:
Fixes #14332

sdmaclea committed:
[Arm64] EXTERNAL_FeatureSIMD_Default = 1

sdmaclea committed:
[Arm64] Enable SIMD compilation

BruceForstall commented on src/jit/CMakeLists.txt:
```diff
@@ -8,6 +8,10 @@ if (CLR_CMAKE_TARGET_ARCH_AMD64 OR (CLR_CMAKE_TARGET_ARCH_I386 AND NOT CLR_CMAKE
 endif ()
 
+if (CLR_CMAKE_TARGET_ARCH_ARM64)
+  add_definitions(-DFEATURE_SIMD)
```
why not add it as another "OR" above?

BruceForstall replied:
You'll also need to alter protononjit\CMakeLists.txt: that file removes FEATURE_SIMD (since x86/x64 define it, but the x86_arm and x64_arm64 altjits currently do not support SIMD). You'll need to move the "remove" into the "I386" section, since the x64_arm64 altjit will support SIMD.

We might need to think about what needs to happen, if anything, to make the altjit work equivalently to the native JIT.

(btw, I presume one of your changes fixes `CILJit::getMaxIntrinsicSIMDVectorLength`)

sdmaclea replied:
> why not add it as another "OR" above?

If I had trouble with windows compilation, I was thinking about disabling, so the logic above would get complex.

I will move it.

> You'll also need to alter protononjit\CMakeLists.txt

OK

> (btw, I presume one of your changes fixes CILJit::getMaxIntrinsicSIMDVectorLength)
I have one `getMaxIntrinsicSIMDVectorLength` change.  `CILJIT` namespace does not look familiar.  I'll take a look when I get back to the office.

BruceForstall commented on src/jit/CMakeLists.txt:
```diff
@@ -138,6 +142,7 @@ set( JIT_ARM64_SOURCES
   lowerarm64.cpp
   lsraarmarch.cpp
   lsraarm64.cpp
+  simd.cpp
```
You also need to add simdcodegenarm64.cpp

sdmaclea replied:
Right now I put all the simdcodegen in codegenarm64.  If it is preferred I can create a separate file.

sdmaclea commented:
@BruceForstall I have handled your comments where they were still relevant.

It may be time to remove the NoMerge tag and start compiling FEATURE_SIMD for ARM64.  It would still be disabled by default.

BruceForstall commented:
@sdmaclea I removed the "no merge" label. We should do a full set of testing, including stress modes, before merging. Unfortunately, our CI system is down, maybe for the weekend.

sdmaclea commented:
test this please

BruceForstall commented:
@dotnet-bot test this please

BruceForstall commented:
@dotnet-bot test Windows_NT x86_arm_altjit Checked Build and Test
@dotnet-bot test Windows_NT x64_arm64_altjit Checked Build and Test

BruceForstall commented:
@dotnet-bot test this please

BruceForstall commented:
@dotnet-bot test Windows_NT x86_arm_altjit Checked Build and Test
@dotnet-bot test Windows_NT x64_arm64_altjit Checked Build and Test

sdmaclea commented:
@BruceForstall Do you expect the altjit tests to pass on arm64?  Is SIMD enabled in altjit because it is enabled by default on x64?  I wouldn't expect all the tests to pass with SIMD enabled (yet).

BruceForstall commented:
> @BruceForstall Do you expect the altjit tests to pass on arm64?

yes. Currently, the x64_arm64_altjit jobs are almost clean, and we want to keep them that way.

You enabled FEATURE_SIMD in protononjit\CMakeLists.txt.

> Is SIMD enabled in altjit because it is enabled by default on x64?

It's possible that since x64 in SIMD is enabled, the VM is telling the altjit for arm64 that it is enabled. Perhaps we need to temporarily add some code in the JIT under `defined(ALT_JIT) && defined(_TARGET_ARM64_)` to stop this. Maybe here:

```
#ifdef FEATURE_SIMD
    // Minimum bar for availing SIMD benefits is SSE2 on AMD64/x86.
    featureSIMD = jitFlags->IsSet(JitFlags::JIT_FLAG_FEATURE_SIMD);
    setUsesSIMDTypes(false);
#endif // FEATURE_SIMD
```

BruceForstall commented:
Or, simply don't enable FEATURE_SIMD for `_TARGET_ARM64_` in protononjit\CMakeLists.txt until it is ready.

BruceForstall commented:
@dotnet-bot test this please

BruceForstall commented:
@dotnet-bot test Windows_NT x86_arm_altjit Checked Build and Test
@dotnet-bot test Windows_NT x64_arm64_altjit Checked Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress2 Build and Test

BruceForstall commented:
@dotnet-bot test Windows_NT arm64 Cross Debug Innerloop Build

sdmaclea commented:
@BruceForstall Looks like I need to figure out how to run arm64 altjit on Ubuntu x64 to debug these fails.

BruceForstall commented:
> @BruceForstall Looks like I need to figure out how to run arm64 altjit on Ubuntu x64 to debug these fails.

protononjit is already building on Ubuntu x64. You should be able to:
```
export COMPlus_AltJit=*
export COMPlus_AltJitNgen=*
export COMPlus_AltJitName=libprotononjit.so
```
then run the test in question.

sdmaclea commented:
The altjit issue is that https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L1202 is asking the `primary jit` about the vector length instead of using the altjit.

If I add `export COMPlus_EnableAVX=0` the failing test I was debugging compiles correctly when using altjit.

BruceForstall commented:
Interesting; I guess the vector size on x64 is 32 and arm64 is 16, hence the problem. I wonder if we should add code there (in the VM) to ask the altjit, if one is loaded. I guess there's a potential issue here where you can use altjit on only one method, not all, and that might not work well.

@CarolEidt what do you think?

sdmaclea commented:
Looks like the jitstress failures match the baseline.  Probably the issues you already opened.

sdmaclea commented:
> I wonder if we should add code there (in the VM) to ask the altjit, if one is loaded. I guess there's a potential issue here where you can use altjit on only one method, not all, and that might not work well.

I think if we remember the active jit (the one currently compiling) it should work.

BruceForstall commented:
I tried changing `MethodTableBuilder::CheckIfSIMDAndUpdateSize()` to use the altjit if one is loaded, but then after the arm64 altjit compiles, it falls back to the x64 jit and it asserts/fails. I guess this is because the VM builds the type information just once, and needs the correct vector size to make that happen. This is one of those cases where the altjit model doesn't work well, whereas the true "crossgen" model does.

It might be best for now to not enable ARM64 SIMD in protononjit, unless someone else has a good idea for making it work.

sdmaclea commented:
> ideas

Set CI to `export COMPlus_EnableAVX=0` for `arm64_altjit` runs?

sdmaclea commented:
@BruceForstall disabled SIMD in altjit for now

CarolEidt commented:
> Interesting; I guess the vector size on x64 is 32 and arm64 is 16, hence the problem. I wonder if we should add code there (in the VM) to ask the altjit, if one is loaded. I guess there's a potential issue here where you can use altjit on only one method, not all, and that might not work well.

I think the right way to support the altjit would be to effectively constrain the vector size to 16 (i.e. as if `COMPlus_EnableAVX == 0`), possibly in `CILJit::getMaxIntrinsicSIMDVectorLength()`.

If we don't want to enable `FEATURE_SIMD` yet for arm64, it would at least be good to have a way to turn it on for testing, but I'm not sure, off the top of my head, what the best way to do that would be.

sdmaclea committed:
[Arm64] Allow SIMD in altjit

Add flag SIMD16ByteOnly
Set COMPlus_SIMD16ByteOnly=1 in x64_arm64_altjit.cmd
If SIMD16ByteOnly, limit clear FLAG_USE_AVX2 to disable SIMD32 vector size
Enable SIMD in protonjit

sdmaclea committed:
[Arm64] Enable SIMD in crossgen

sdmaclea commented:
@dotnet-bot test Windows_NT x86_arm_altjit Checked Build and Test
@dotnet-bot test Windows_NT x64_arm64_altjit Checked Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress2 Build and Test

sdmaclea commented:
@BruceForstall @CarolEidt 

+ I tried to implement Carol's suggestion by introducing a new config variable SIMD16ByteOnly
+ Re-enabled SIMD in altjit
+ I also combined this with #14634, which is also ready.
+ I enabled SIMD in crossgen per the suggestion in #14634.

sdmaclea commented:
@dotnet-bot test Windows_NT x86_arm_altjit Checked Build and Test
@dotnet-bot test Windows_NT x64_arm64_altjit Checked Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress2 Build and Test

sdmaclea commented:
@dotnet-bot test Windows_NT x86_arm_altjit Checked Build and Test
@dotnet-bot test Windows_NT x64_arm64_altjit Checked Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress2 Build and Test

sdmaclea commented:
Looks like my altjit attempt failed.  I will take another look at fixing it.

sdmaclea commented:
There is one SIMD JitStress failure I am trying to debug before I try to fix altjit again

sdmaclea commented:
OK, JitStress=1/2 is now clean on my local tip.  I have uploaded PR's for all JitStress issues found.  I will make another attempt to fix altjit.

sdmaclea commented:
I have simplified the x64_arm64_altjit to simply Clear FLAG_USE_AVX2.  This is the simplest way to disable SIMD32 support during altjit runs.  It will disable AVX2 everywhere (SIMD and intrinsics) when running arm64 altjit, but since that is still a supported configuration, it seems OK.  I kept the new config variable name.

sdmaclea commented:
@dotnet-bot test Windows_NT x64_arm64_altjit Checked Build and Test

I will not trigger the JitStress jobs until the PR to fix merge

CarolEidt commented on src/zap/zapper.cpp:
```diff
@@ -1394,6 +1394,14 @@ void Zapper::InitializeCompilerFlags(CORCOMPILE_VERSION_INFO * pVersionInfo)
+        m_pOpt->m_compilerFlags.Set(CORJIT_FLAGS::CORJIT_FLAG_FEATURE_SIMD);
+    }
+#endif
+
```
Why is this code needed? I may be mistaken, but I don't believe this will impact SIMD code generation, because the JIT will reject any pre-jitting of an method that has SIMD in it.

sdmaclea replied:
@CarolEidt : @BruceForstall Suggested enabling SIMD code generation during crossgen since it is currently supported on all arm64 platforms.  This was my attempt to do that.  I thought this would be sufficient...

CarolEidt replied:
I'm having some difficulty tracking down how code generation is disabled for `Vector<T>`. I thought there was a check in compiler.cpp, but can't find it. For x86/x64, it is important to not simply disable FeatureSIMD, but to ensure that no pre-compiled code is generated for a method that uses `Vector<T>` since its size is not known until runtime. I believe that check needs to be done in the JIT, so I was looking for where that happens.

sdmaclea replied:
@CarolEidt 

So I just verified crossgen is generating SIMD enabled code for ARM64.  Code looks mostly identical except linkage related changes.

I think precompiling `Vector<T> ` is OK, because the importer replaces the SIMD Vector<T> method calls with intrinsics.  However, running x64 crossgen on a SIMD benchmark is probably a bad idea.  As I see do not see a trap on method that might have SIMD.  In fact since featureSIMD is disabled during crossgen it will not even be detected.

sdmaclea replied:
Opened #15328

sdmaclea replied:
I believe this discussion was resolved/closed in #15328

CarolEidt reviewed (commented):
The approach LGTM - @BruceForstall what do you think?
I'm unsure about what the zapper change is for.

sdmaclea commented:
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress2 Build and Test

sdmaclea commented:
test Windows_NT arm64 Cross Checked Innerloop Build and Test

sdmaclea commented on src/inc/clrconfigvalues.h:
```diff
@@ -529,14 +529,18 @@ CONFIG_DWORD_INFO_EX(INTERNAL_JitLoopHoistStats, W("JitLoopHoistStats"), 0, "Dis
+#else // !(defined(_TARGET_AMD64_) || defined(_TARGET_X86_) || defined(_TARGET_ARM64_))
 #define EXTERNAL_FeatureSIMD_Default 0
+#endif // !(defined(_TARGET_AMD64_) || defined(_TARGET_X86_) || defined(_TARGET_ARM64_))
+#if defined(_TARGET_XARCH_)
```
@BruceForstall Does this need to be `#if defined(_TARGET_AMD64_) || defined(_TARGET_X86_)` ?

BruceForstall replied:
yes.

CarolEidt replied:
I'm confused how does:
```
#if defined(_TARGET_AMD64_) || defined(_TARGET_X86_)
```
differ from what's there:
```
#if defined(_TARGET_XARCH_)
```

sdmaclea replied:
In previous changes @BruceForstall has indicated that `_TARGET_XARCH_` should not be used in certain files.

CarolEidt replied:
Ah I see - thanks for the clarification.

BruceForstall replied:
In particular, it's only defined for the JIT, and clrconfigvalues.h is included by the VM.

sdmaclea replied:
Done

CarolEidt reviewed (commented):
Still looks good - not sure about the `#if`s

sdmaclea committed:
Fix #if per feedback

sdmaclea commented:
@dotnet-bot test Windows_NT x64_arm64_altjit Checked Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress1 Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked jitstress2 Build and Test

sdmaclea commented:
Only `\tighttimercallback` failed in both jitstress runs.  This is a existing issue and can be ignored for this PR.

I think this is ready to merge

CarolEidt commented:
@BruceForstall are there more changes you would like to see? This LGTM

BruceForstall reviewed (approved):
Thanks for doing the work to keep altjit working!

BruceForstall commented:
@sdmaclea Thank you for all your work to get ARM64 SIMD implemented!

xiangzhai commented:
:mips-interest
